storage/range.go

```diff
@@ -1629,28 +1629,84 @@ func (r *Range) FirstIndex() (uint64, error) {
 
 // Snapshot implements the raft.Storage interface.
 func (r *Range) Snapshot() (raftpb.Snapshot, error) {
-	// TODO(bdarnell): package up all the data in the range, not just
-	// the descriptor (and send the appropriate log index).
-	data, err := gogoproto.Marshal(r.Desc())
+	// Copy all the data from a consistent RocksDB snapshot into a RaftSnapshotData.
+	snap := r.rm.NewSnapshot()
+	defer snap.Stop()
+	var snapData proto.RaftSnapshotData
+
+	// Certain keys are treated specially while building the snapshot;
+	// see the switch statement below for details.
+	hardStateKey := engine.RaftHardStateKey(r.Desc().RaftID)
```

Comment:
I would recommend the following changes:

- In the ApplySnapshot code, first read existing value of the hard state key, then apply the snapshot data in full, then rewrite the existing value of the hard state key. This will save a lot of CPU on snapshots, as we won't be decoding the keys at each step. It will also make this method a lot simpler.

- Use MVCCGet() for reading the applied index.

- Use MVCCGet() with consistent=false for reading the descriptor. We must leave the intents for the descriptor exactly as is. This is a tricky part. Keep in mind that there's no way to tell whether the intent's value for the descriptor is correct or the non-intent value. The correct value for the descriptor is only ascertainable by resolving the intent. So we can't use this value! However, the value we can use is the value of the descriptor that this replica holds in memory. Afterall, if the transaction was completed, that value was set in the update replicas trigger. If not, then this replica's descriptor is correct as well--it's the pre-intent value and will be until the transaction is completed.